runtime/onert/core/src/exporter/CircleExporter.cc

```diff
@@ -81,6 +83,50 @@ void CircleExporter::updateWeight(const std::unique_ptr<exec::Execution> &exec)
     });
 }
 
+void CircleExporter::updateMetadata(const std::unique_ptr<ir::train::TrainingInfo> &training_info)
+{
+  const char *const TRAININFO_METADATA_NAME = "CIRCLE_TRAINING";
+
+  TrainInfoBuilder tbuilder(training_info);
+  static std::mutex mutex;
+
+  bool found = false;
```

Comment:
code.i's suggestion,
```suggestion
  std::atomic<bool> found(false);
```

Reply:
Here is the overall explanation about this by "code.i":

---

This code is part of the `CircleExporter` class, which updates the metadata of a `Circle` model using a `TrainingInfo` object. This code has two main issues:

Race condition: The code uses a `static std::mutex` to protect against concurrent access to the metadata while updating it, but it also uses a `found` variable to check if the metadata has been found within the `for` loop. This can lead to race conditions where multiple threads try to change the `found` variable at the same time. To fix this, we can change the `found` variable to be an `std::atomic<bool>`.

Memory leak: The code copies the `TrainingInfo` object into the metadata of the `Circle` model using `memcpy()`, but it does not release the `TrainingInfo` object. To fix this, we can wrap the `TrainingInfo` object in a smart pointer and let it automatically release its memory when it goes out of scope.

By fixing these issues, we can ensure that the code is thread-safe and avoids memory leaks.

---

Reply:
Anyway, are you planning to call this function with multi-threading? If so, several more variables need to be considered for concurrency,, although code.i does not mentioned them.

Reply:
I did not design this CircleExporter class considering multi-thrading. However, the race condition should be considered when operating the buffer, so Mutex was added. I update this code according to your suggestion. Thank you.

Reply:
Sorry for confusion. I guess this variable is a local variable, and local variable is thread-safe.
I think "code.i" informs a wrong information. This variable had no problem.